### DISCUSSION TITLE
OpenFilePicker crashes.

### DISCUSSION THREAD

**Discussion by garawaa** [unoplatform/uno#10875] [Q&A]
```
var fileOpenPicker = new FileOpenPicker();
fileOpenPicker.SuggestedStartLocation = PickerLocationId.ComputerFolder;
fileOpenPicker.FileTypeFilter.Add(".txt");
fileOpenPicker.FileTypeFilter.Add(".csv");
StorageFile pickedFile = await fileOpenPicker.PickSingleFileAsync();
if (pickedFile != null)
{
    // File was picked, you can now use it
}
else
{
    // No file was picked or the dialog was cancelled.
}
```
I just copied this code from uno doc. Compiles without any error but when attemp to open a dialog app just crashes.
OS windows 11 home edition. tool: vs 2022 community preview.  All uno-check completed.
Exception details:
Unhandled exception at 0x00007FFD578D0F2C (CoreMessagingXP.dll) in UnoApp1.Windows.exe: 0xC000027B: An application-internal exception has occurred (parameters: 0x000001DA042B42B0, 0x0000000000000001).
Unhandled exception at 0x00007FFE3F53E1EF (combase.dll) in UnoApp1.Windows.exe: 0xC0000602:  A fail fast exception occurred. Exception handlers will not be invoked and the process will be terminated immediately.

**Comment by FrozDark** [comment]
The Windows build is handled by WinUI team themself. Ask questions related to it there https://github.com/microsoft/microsoft-ui-xaml

As for the issue, you have to initialize a picker with the current window handle as asnwered here https://github.com/microsoft/WindowsAppSDK/issues/1188

**Reply by garawaa** [reply]
Above code works fine for UWP project. But does not work under uno platform.

**Reply by francoistanguay** [reply]
THe crash is on the Windows head? If so, Uno isn't present. If your head is WinUI, then what @FrozDark  mentioned is correct.

**Comment by MartinZikmund** [comment]
@garawaa This is not UWP, but rather Windows App SDK (judging by the `.Windows` suffix). The reason this crashes is that for Windows App SDK Microsoft now requires two additional lines which were not needed in UWP (it is due to the fact that WinAppSDK allows for multiple windows at once). Add the following lines before the call to `PickSingleFileAsync`:

```csharp
var hwnd = WinRT.Interop.WindowNative.GetWindowHandle(window);
WinRT.Interop.InitializeWithWindow.Initialize(fileOpenPicker, hwnd);
```

Where `window` is the main window of the application. The easiest way to get it is to make the `_window` filed in `App.xaml.cs` `internal static` and then use `App._window` as the argument.

**Reply by sasakrsmanovic** [reply]
@garawaa - Are you able to provide the link to the doc you had used so we can adjust that doc if needed please?

**Reply by MartinZikmund** [reply]
I have updated the docs in a PR that was merged yesterday, so they will soon be up to date with this

**Reply by garawaa** [reply]
```csharp
var fileOpenPicker = new FileOpenPicker();
fileOpenPicker.SuggestedStartLocation = PickerLocationId.ComputerFolder;
fileOpenPicker.FileTypeFilter.Add(".txt");
fileOpenPicker.FileTypeFilter.Add(".csv");
#if !HAS_UNO
var hwnd = WinRT.Interop.WindowNative.GetWindowHandle(window);
WinRT.Interop.InitializeWithWindow.Initialize(fileOpenPicker, hwnd);
#endif
StorageFile pickedFile = await fileOpenPicker.PickSingleFileAsync();
if (pickedFile != null)
{
    // File was picked, you can now use it
}
else
{
    // No file was picked or the dialog was cancelled.
}
```
with it still gives error in windows.

**Reply by garawaa** [reply]
[code doc link](https://platform.uno/docs/articles/features/windows-storage-pickers.html)

**Reply by jeromelaban** [reply]
[Here's the updated document](https://github.com/MartinZikmund/Uno/blob/master/doc/articles/features/windows-storage-pickers.md) that @MartinZikmund was talking about.

**Reply by MartinZikmund** [reply]
@garawaa Is the updated code working? What error are you getting on Windows?

**Reply by garawaa** [reply]
No sorry.

**Reply by Youssef1313** [reply]
@MartinZikmund `#if !HAS_UNO` won't work if the Windows project is referencing `Uno.WinUI`, in which case `HAS_UNO` is going to be `true` and the conditional code will not take effect.

@garawaa Consider adding the following property group to the `YourApp.Windows.csproj`:

```xml
<PropertyGroup>
    <DefineConstants>$(DefineConstants);IS_WINUI</DefineConstants>
</PropertyGroup>
```

and do the suggestion from Martin but now using `#if IS_WINUI` instead of `#if !HAS_UNO`

**Reply by MartinZikmund** [reply]
@garawaa Actually you can now completely drop the `#if`, as it is no longer needed, the `Interop` namespace is available in Uno as well

@Youssef1313 we should adjust this definitely 🤔

**Reply by Youssef1313** [reply]
Sorry, my bad. `HAS_UNO` isn't defined in Windows heads.

**Reply by Youssef1313** [reply]
@garawaa If it's not working with Martin suggestions, please upload a repro project so we can take a look.

**Reply by MartinZikmund** [reply]
Feel free to create an issue (bug) for better tracking.

**Comment by garawaa** [comment]
[Screen record](https://imgur.com/a/xLyaEx3)
